rule/package-comments.go

```diff
@@ -31,9 +31,48 @@ func (*PackageCommentsRule) Apply(file *lint.File, _ lint.Arguments) []lint.Fail
 	fileAst := file.AST
 	w := &lintPackageComments{fileAst, file, onFailure}
 	ast.Walk(w, fileAst)
+	failures = append(failures, checkPackageComment(file.Pkg)...)
 	return failures
 }
 
+var checkPackageCommentCache = map[*lint.Package]struct{}{}
+
+func checkPackageComment(p *lint.Package) []lint.Failure {
+	// deduplicate warnings in package
+	if _, ok := checkPackageCommentCache[p]; ok {
+		return nil
+	}
+	checkPackageCommentCache[p] = struct{}{}
+	var docFile *ast.File
+	var firstFile *ast.File
```

Comment:
what is the purpose of these `first*` variables? _first_ in which sense?

Reply:
for a package with many source file, and all of them have no package doc.
I think we should only raise one failure to warning lacking of pacakge doc
rather than warning on every file. (so there is checkPackageCommentCache)
so the problem is warning on which file.
currently I perfer warning on `doc.go` if it is exists.
else warning on then `firstFile` which is the first file in lexicographic order.

Reply:
@okhowang, can we also add a special case for files named after the package in addition to `doc.go`? E.g. it’s common to have package `foo` docs in `foo.go`.

Reply:
> @okhowang, can we also add a special case for files named after the package in addition to `doc.go`? E.g. it’s common to have package `foo` docs in `foo.go`.

of course